RecoLocalTracker/SiStripClusterizer/plugins/SiStripRawToClusterGPUKernel.cc

```diff
@@ -0,0 +1,184 @@
+  }
+
+  SiStripRawToClusterGPUKernel::SiStripRawToClusterGPUKernel(const edm::ParameterSet& conf)
+      : fedIndex_(sistrip::NUMBER_OF_FEDS, stripgpu::invalidFed),
+        channelThreshold_(conf.getParameter<double>("ChannelThreshold")),
+        seedThreshold_(conf.getParameter<double>("SeedThreshold")),
+        clusterThresholdSquared_(std::pow(conf.getParameter<double>("ClusterThreshold"), 2.0f)),
+        maxSequentialHoles_(conf.getParameter<unsigned>("MaxSequentialHoles")),
+        maxSequentialBad_(conf.getParameter<unsigned>("MaxSequentialBad")),
+        maxAdjacentBad_(conf.getParameter<unsigned>("MaxAdjacentBad")),
+        maxClusterSize_(conf.getParameter<unsigned>("MaxClusterSize")),
+        minGoodCharge_(clusterChargeCut(conf)) {
+    fedRawDataOffsets_.reserve(sistrip::NUMBER_OF_FEDS);
+  }
+
+  void SiStripRawToClusterGPUKernel::makeAsync(const std::vector<const FEDRawData*>& rawdata,
+                                               const std::vector<std::unique_ptr<sistrip::FEDBuffer>>& buffers,
+                                               const SiStripClusterizerConditionsGPU& conditions,
+                                               cudaStream_t stream) {
+    size_t totalSize{0};
+    for (const auto& buff : buffers) {
+      if (buff != nullptr) {
+        totalSize += buff->bufferSize();
+      }
+    }
+
+    fedRawDataHost_ = cms::cuda::make_host_unique<uint8_t[]>(totalSize, stream);
+    auto fedRawDataGPU = cms::cuda::make_device_unique<uint8_t[]>(totalSize, stream);
+
+    size_t off = 0;
+    fedRawDataOffsets_.clear();
+    fedIndex_.clear();
+    fedIndex_.resize(sistrip::NUMBER_OF_FEDS, stripgpu::invalidFed);
+
+    sistrip::FEDReadoutMode mode = sistrip::READOUT_MODE_INVALID;
+
+    for (size_t fedi = 0; fedi < buffers.size(); ++fedi) {
+      auto& buff = buffers[fedi];
+      if (buff != nullptr) {
+        const auto raw = rawdata[fedi];
+        memcpy(fedRawDataHost_.get() + off, raw->data(), raw->size());
+        fedIndex_[stripgpu::fedIndex(fedi)] = fedRawDataOffsets_.size();
+        fedRawDataOffsets_.push_back(off);
+        off += raw->size();
+        if (fedRawDataOffsets_.size() == 1) {
+          mode = buff->readoutMode();
+        } else {
+          if (buff->readoutMode() != mode) {
+            throw cms::Exception("[SiStripRawToClusterGPUKernel] inconsistent readout mode ")
+                << buff->readoutMode() << " != " << mode;
+          }
+        }
+      }
+    }
+    // send rawdata to GPU
+    cms::cuda::copyAsync(fedRawDataGPU, fedRawDataHost_, totalSize, stream);
+
+    const auto& detmap = conditions.detToFeds();
+    if ((mode != sistrip::READOUT_MODE_ZERO_SUPPRESSED) && (mode != sistrip::READOUT_MODE_ZERO_SUPPRESSED_LITE10)) {
+      throw cms::Exception("[SiStripRawToClusterGPUKernel] unsupported readout mode ") << mode;
+    }
+    const uint16_t headerlen = mode == sistrip::READOUT_MODE_ZERO_SUPPRESSED ? 7 : 2;
```

Comment:
should these magic numbers be constexpr somewhere?

Reply:
to date in cmssw, this logic is done in the same way. did you have in mind adding two constexprs here? (eg, to improve readability)